src/core.rs

```diff
@@ -730,20 +990,64 @@ pub fn wrap_optimal_fit<'a, T: Fragment, F: Fn(usize) -> usize>(
         cost
     });
 
-    let mut lines = Vec::with_capacity(line_numbers.get(fragments.len(), &minima));
-    let mut pos = fragments.len();
-    loop {
-        let prev = minima[pos].0;
-        lines.push(&fragments[prev..pos]);
-        pos = prev;
-        if pos == 0 {
-            break;
+    let mut line_start = minima[fragments.len()].0;
+
+    WrapOptimalFit {
+        fragments: fragments
+            .into_iter()
+            .enumerate()
+            .rev()
+            .map(|(i, data)| {
+                let eol = i + 1 == line_start;
+                if eol {
+                    line_start = minima[line_start].0;
+                }
+                (data.fragment, eol)
+            })
```

Comment:
Have you noticed how this changes the complexity of this piece of code to O(`number_of_fragments`)? Before it was O(`number_of_lines`). Because `wrap_optimal_fit` returns an iterator which yields every single `Fragment`, this seems to be built into the design — which is unfortunate. It feels like a "tax" that we should be able to avoid.

Could `wrap_optimal_fit` and `wrap_first_fit` instead return an `Iterator<Item=Line>` where `Line` is itself something which encapsulates, well, a line of text. This would basically mean grouping the `Fragment`s according to the EOL `bool` you have, but the rest of the layers would no longer deal with individual `Fragment`s but instead deal in full lines. Similar to how the upper layers deal in `&[Fragment]` today.

Reply:
The problem with a `Line` type is that the complexity within the function would still be O(`number_of_fragments`) as each fragment would have to be copied into the line, and it would incur a fairly large allocation cost - each line would probably need a `Vec`.

Almost all code that uses the returned iterator currently iterates over each fragment in the line, so the overall complexity would remain the same most of the time. This approach also avoids double-loops, which _might_ be faster overall, but I'm not sure.